app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java

```diff
@@ -0,0 +1,283 @@
+                                finish();
+                            }
+                        }
+                );
+        dialogBuilder.show();
+    }
+
+    /**
+     * Shows an alert dialog about the status of the print job.
+     * Activity will quit upon exiting the dialog.
+     *
+     * @param msg String message that should be shown on the alert
+     */
+    public void showAlertDialog(String msg) {
+        AlertDialog.Builder alert = new AlertDialog.Builder(this);
+        alert.setTitle(msg);
+        alert.setCancelable(false);
+        alert.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
+
+            public void onClick(DialogInterface dialog, int which) {
+                dialog.dismiss();
+                finish();
+            }
+        });
+
+        alert.show();
+    }
+
+    /**
+     * Prepares a WebView of the html document generated by this activity, which can then be
+     * printed by the Android print framework
+     *
+     * Source: https://developer.android.com/training/printing/html-docs.html
+     */
+    private void doHtmlPrint() {
+        // Create a WebView object specifically for printing
+        WebView webView = new WebView(this);
+        webView.setWebViewClient(new WebViewClient() {
+
+            public boolean shouldOverrideUrlLoading(WebView view, String url) {
+                return false;
+            }
+
+            @Override
+            public void onPageFinished(WebView view, String url) {
+                createWebPrintJob(view);
+                mWebView = null;
+            }
+        });
+        try {
+            String htmlDocString = TemplatePrinterIOUtil.readStringFromFile(outputPath);
+            webView.loadDataWithBaseURL(null, htmlDocString, "text/HTML", "UTF-8", null);
+
+            // Keep reference to WebView object until PrintDocumentAdapter is passed to PrintManager
+            mWebView = webView;
+        } catch (IOException e) {
+            showErrorDialog("Could not read from generated HTML doc in order to print");
+        }
+
+    }
+
+    /**
+     * Starts a print job for the given WebView
+     *
+     * Source: https://developer.android.com/training/printing/html-docs.html
+     *
+     * @param v the WebView to be printed
+     */
+    @TargetApi(Build.VERSION_CODES.KITKAT)
+    private void createWebPrintJob(WebView v) {
+
+        // Get a PrintManager instance
+        PrintManager printManager = (PrintManager) getSystemService(Context.PRINT_SERVICE);
+
+        // Get a print adapter instance
+        PrintDocumentAdapter printAdapter = v.createPrintDocumentAdapter();
+
+        // Create a print job with name and adapter instance
+        printJob = printManager.print(jobName, printAdapter,
+                new PrintAttributes.Builder().build());
+        jobStarted = true;
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        if (jobStarted) {
```

Comment:
I'll just say this preemptively- I know it's kind of hacky to have a check in onResume like this, but it was the only way I could come up with for knowing when to finish the activity. Unlike when we were printing a PDF and creating a custom PrintDocumentAdapter for it, a WebView (very conveniently) has a built-in method for creating its own adapter (SEE v.createPrintDocumentAdapter() above). However, this removes the opportunity to place any sort of listener for the result of the job within the onFinish() method of the adapter. 

To give some context of why this works and why I'm confident it doesn't break anything- when the print job is started in createWebPrintJob(), an external Android print dialog pops up where the user can choose what printer to print to, save to PDF, etc. At that point, a return to the print activity can be due to only one of 2 things: 1) the print dialog closes naturally, signaling the print job has ended, or 2) the user presses back. In either case, the desired behavior at that point is for the activity to end (after telling the user what happened with the alert dialog). Technically, there is a 3rd way to end up back at the print activity, and that would be to rotate the screen. However, since this causes the activity to start up entirely new anyway (from onCreate() rather than on Resume()), this has no effect.

All that being said, if anyone has a cleaner idea for how to handle this, let me know!